Land grab in space

colinleach opened the pull request:
Pairs with `composite-types`.

Based on the C# version with the same name, but I ended up making lots of changes.

colinleach commented:
It looks like the previous test failures were with the result vector being (randomly) out of order. As an easy fix, I switched to returning a set instead of a vector.

depial commented on exercises/concept/land-grab-in-space/.docs/instructions.md:
Typo: "agains" => "against"

depial commented on exercises/concept/land-grab-in-space/.docs/instructions.md:
I think both options should be returned as a "register" (i.e. as a set of `Plot`s) for consistency's sake.

depial commented on exercises/concept/land-grab-in-space/.meta/exemplar.jl:
```diff
@@ -0,0 +1,30 @@
+    top_right::Coord
+end
+
+is_claim_staked(claim::Plot, register::Set{Plot}) = claim ∈ register
```
I believe this may be too simplistic given the story (see comments about tests).

depial commented on exercises/concept/land-grab-in-space/runtests.jl:
```diff
@@ -0,0 +1,84 @@
+            @test is_claim_staked(plot, register)
+        end
+
+        @testset "does not exist" begin
+            plot1 = Plot(bottom_left=Coord(1, 1), top_right=Coord(2, 3))
+            plot2 = Plot(bottom_left=Coord(1, 0), top_right=Coord(2, 5))
+            register = Set{Plot}([plot1,])
+            @test !is_claim_staked(plot2, register)
+        end
```
`plot1` and `plot2` are overlapping (indeed `plot1` is a strict subset of `plot2`), so I don't think `plot2`'s claim should be allowed. To fix this, I guess we should probably change `plot2` to something non-overlapping with `plot1`.

Originally, this test made me think that the claims could be arbitrary, rather than from a predefined set of plots, so I solved it for the former (hence my comment about the solution being "simplistic"). However, I see that the instructions specify the latter, so you can ignore that comment. However, we may want to revisit the wording of that part of the instructions to avoid confusion of this type.

colinleach replied:
Oops! Brain failure...

I've also made some changes at the top of the instructions to clarify that plots are non-overlapping (and by implication, the tests can be simplistic).

depial replied:
These changes look good!

depial commented on exercises/concept/land-grab-in-space/.docs/instructions.md:
This could be:

"Speculators claim one of the plots of land by specifying its coordinates."

depial commented on exercises/concept/land-grab-in-space/.docs/hints.md:
```diff
@@ -0,0 +1,56 @@
+- You already wrote a function to calculate the longest side of a plot, and can apply this to the whole collection.
+- See Task 5 for a note on `max` versus `maximum`.
+- Find one or more plots with the maximum length.
+- Return either a `Plot` or a `Vector{Plot}` as appropriate.
```
This will need to be brought in line with the end testing decision (currently `Vector{Plot}` => `Set{Plot}`)

colinleach commented:
I think I've now addressed all your points, though not at the first attempt.

I'll change this to "Ready for review", though there's no rush to approve it if you're still traveling. It's still wip, so invisible to students. We don't want unconnected concepts floating around now the syllabus is live.

depial commented on exercises/concept/land-grab-in-space/.docs/hints.md:
````diff
@@ -0,0 +1,56 @@
+3
+```
+
+## 6. Find the plot claimed that has the longest side, for research purposes.
+
+- The register is a Set, unordered but iterable.
+- It may be convenient to convert the register to something ordered, for example with `collect()`.
+- You already wrote a function to calculate the longest side of a plot, and can apply this to the whole collection.
+- See Task 5 for a note on `max` versus `maximum`.
+- Find one or more plots with the maximum length.
+- Return the results as a `Set{Plot}`.
````
On first inspection, this struck me as being potentially overly specific to the `exemplar.jl` solution, but I'm not sure if that is intended/desired. To be clear, I quite like your `exemplar.jl` solution because it shows off a lot of idiomatic Julia, so I don't mind this pushing that direction. 

For reference, when I first solved this, my main aim was to only loop over the register once, so it turned out a bit more generic:

```julia
function get_claim_with_longest_side(register::Set{Plot})::Set{Plot}
    longestsides, longestside = Set(), 0
    for claim in register
        claimmax = get_longest_side(claim)
        claimmax == longestside && push!(longestsides, claim)
        claimmax > longestside && (longestsides = Set([claim]); longestside = claimmax)
    end
    longestsides
end
```

To be fair, if we were to make the hints more generic, I'm not sure exactly how to go about it... A first thought is that I could see putting point 2 and point 3 into a sub-list under point 1 (with slightly more open wording) possibly?

- The register is a Set, unordered but iterable.
  - It may be convenient to convert the register to something ordered, for example with `collect()`.
  - You already wrote a function to calculate the longest side of a plot and could apply this to the whole collection.
- See Task 5 for a note on `max` versus `maximum`.
- Find one or more plots with the maximum length.
- Return the results as a `Set{Plot}`.

Again, I'm fine with leaving the hint as is, but I thought I'd at least mention this.

colinleach replied:
Thanks for the suggestions. I'll have a think about this tomorrow, when I hope to be thinking more clearly.

colinleach replied:
Getting to the point of "thinking more clearly" has been slower than I hoped!

For now, I've made the changes you suggested and will merge as wip. We can change it later if we get any better ideas.

depial reviewed (approved):
This is looking good! Just one more niggle to discuss, but it could be ignored if you'd like.
